Improve parsing of HTTP_HOST header

pascalbetz opened the pull request:
IPV6 Host was not properly parsed.

With the insanely detailed bug report his was fixed in no time:  👍



Closes #2584

### Description
Please describe your pull request. Thank you for contributing! You're the best.

### Your checklist for this pull request
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed, including Rubocop.

pascalbetz commented on lib/puma/request.rb:
```diff
@@ -231,7 +231,7 @@ def fetch_status_code(status)
     def normalize_env(env, client)
       if host = env[HTTP_HOST]
-        if colon = host.index(":")
+        if colon = host.index(/:\d+$/)
```
Does it make sense to extract this to a constant?

kuei0221 replied:
I have run the benchmark `hello.sh` with your changes, and it seems that extracting to a constant will make it faster. You can give it a try if that makes sense to you as well 😃

No constant
```
❯ benchmarks/wrk/hello.sh
Puma starting in single mode...
* Puma version: 5.2.2 (ruby 2.7.0-p0) ("Fettisdagsbulle")
*  Min threads: 4
*  Max threads: 4
*  Environment: development
*          PID: 23551
* Listening on http://0.0.0.0:9292
Use Ctrl-C to stop
Running 30s test @ http://localhost:9292
  2 threads and 4 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   699.53us  301.57us   4.14ms   81.63%
    Req/Sec     2.90k   622.92     3.97k    66.45%
  Latency Distribution
     50%  612.00us
     75%  813.00us
     90%    1.07ms
     99%    1.81ms
  173587 requests in 30.10s, 10.59MB read
Requests/sec:   5766.76
Transfer/sec:    360.42KB
```

Extract to constant
```
❯ benchmarks/wrk/hello.sh
Puma starting in single mode...
* Puma version: 5.2.2 (ruby 2.7.0-p0) ("Fettisdagsbulle")
*  Min threads: 4
*  Max threads: 4
*  Environment: development
*          PID: 23609
* Listening on http://0.0.0.0:9292
Use Ctrl-C to stop
Running 30s test @ http://localhost:9292
  2 threads and 4 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   623.64us  271.01us  15.81ms   87.05%
    Req/Sec     3.24k   405.07     4.22k    74.09%
  Latency Distribution
     50%  570.00us
     75%  688.00us
     90%    0.89ms
     99%    1.49ms
  194270 requests in 30.10s, 11.86MB read
Requests/sec:   6454.13
Transfer/sec:    403.38KB
```

pascalbetz replied:
@kuei0221 

Thanks for the benchmark results. I tried to reproduce the results and could not find such a difference between requests/s of inline regexp and constant regexp. There was a small difference between the old (`index(":")`) and new version).

I'll try with different ruby version later today.

kuei0221 replied:
You can ignore the benchmark if that doesn't make sense to you ( it defer from machine to machine, and I believe my machine is not very good 😂 )
I think it is a good idea to make it a constant in `const.rb`, so we don't need to allocated this value everytime it is called.
The following comment in that file give us a good reason to do so:
https://github.com/puma/puma/blob/a717b27354011c9f0877e614177b5fba5fc08004/lib/puma/const.rb#L90-L100

pascalbetz commented on test/test_puma_server.rb:
```diff
@@ -53,6 +53,36 @@ def new_connection
     TCPSocket.new(@host, @port).tap {|sock| @ios << sock}
   end
 
+  def test_normalize_host_header_missing
```
There is probably a smarter way to check those headers?

MSP-Greg commented:
A few things:

1.  The comment about '3%' was committed fourteen years ago.
2.  If this is changed to a constant, it could also be placed in the file that calls it.
3.  Using `rindex` instead of `index` would also solve the bug that this is fixing.
4.  Generally, how to handle constants that are only called by one file/class/module is a matter of style.  Often lookup values are treated differently than objects used for code control (branching, conditionals, etc).

pascalbetz commented:
Thanks for the feedback, I'll look into it.
I just realized that my testdata is faulty (missing brackets for ipv6).

I don't think that `rindex` allone would work though. The port is optional in the _host_ header, so an IPV6 without port would return a faulty index.

MSP-Greg commented:
> The port is optional in the host header

Sorry, didn't know/check that...

MSP-Greg commented:
@pascalbetz

Not that familiar with IPv6, couldn't an address without a port also satisfy `/:\d+$/`?

pascalbetz commented:
@MSP-Greg I think my test data for IPV6 is wrong and the Regexp is faulty.

I need to look up the specs to be sure but IMHO the Host header can be

```
<hostname | IPV4 | ("[" IPV6 "]")>
```
along with optional

```
<colon><port>
```
So IPV6 would be

`[3ffe:2a00:100:7031::1]`

and if there is a port

`[3ffe:2a00:100:7031::1]:1234`

If this is correct, then I can write a better RE or do something around `rindex("]:")`

MSP-Greg commented:
I'd vote for `rindex ']:'`

pascalbetz commented:
@MSP-Greg thanks for the feedback. You make it easy for me to contribute.


I added `rindex("]:")` to detect IPV6 with port and added a "not IPV6 check" to the existing logic.

nateberkopec commented on lib/puma/request.rb:
```diff
@@ -231,7 +231,11 @@ def fetch_status_code(status)
     def normalize_env(env, client)
       if host = env[HTTP_HOST]
-        if colon = host.index(":")
+        # host can be a hostname, ipv4 or bracketed ipv6. Followed by an optional port.
```
Good comments 👍

pascalbetz commented:
Hey Puma Team

To me it's easy to contribute to puma. Which is hard work on your end. 

Thanks a lot 🤩

nateberkopec commented:
Thank you for your contribution Pascal!

kaorihinata commented:
I apologize. I was too busy to create the pull request myself at the time. Had I seen this one I would have actually suggested:

```ruby
if host.include?("]:")
  env[SERVER_NAME], _, env[SERVER_PORT] = host.rpartition(":")
elsif not host.include?("[") and host.include?(":")
  env[SERVER_NAME], env[SERVER_PORT] = host.split(":")
else
  env[SERVER_NAME], env[SERVER_PORT] = host, default_server_port(env)
end
```

No offset juggling makes it look less like an awk script, and it's almost identical in performance. On my system it edges out the new code, but only very slightly (hanging in the 458k to 464k range, instead of the 448k to 461k range), but at that point you're averaging `wrk` runs.

I was also going to note that the `hello.sh` script being hard coded to bind to `http://0.0.0.0:9292` is interpreted on some platforms (for example, macOS) to mean "all IPv4 addresses", and `localhost` is always resolved to the IPv4 loopback address, so `wrk` never hits anything beyond the IPv4 address:port scenario.

pascalbetz commented:
oh, did not know `rpartition`

kaorihinata commented:
> oh, did not know `rpartition`

Ruby does have a truly staggering number of methods on some objects.
